### DISCUSSION TITLE
Preventing linkdefinition labels from being lowercased.

### DISCUSSION THREAD

**Discussion by marcofranssen** [remarkjs/remark#601] [Q&A] [answered]
I have the following setup to do some migration of my markdown files from https://hexo.io to https://nextjs.org.

```js
    const contents = remark()
      .use(gfm)
      .use(stringify)
      .use(frontmatter, "yaml")
      .use(fixFrontmatter)
      .use(textr, { plugins: [ellipses] })
      .use(replaceVimeo)
      .use(replaceCodeBlock)
      .use(replaceBlockQuote)
      .use(imgToJsx)
      .use(prefixRelativeLinks, { prefix: "/blog" })
      .processSync(fs.readFileSync(filePath))
      .toString();

    const formatted = prettier.format(contents, {
      ...prettierConfig,
      parser: "mdx",
    });

    fs.writeFileSync(`${outputDir}/${slug}`, formatted);
```

However prettier.format is causing my link labels to be lowercased. Is there any way to keep using prettier, but skip lowercasing the link labels? See below the diff caused by prettier line. I found prettier is using `remark.parse` to format `markdown` and `mdx`.

```diff
-[BlogLetsEncrypt]: /blog/ssl-certificate-for-your-azure-website-using-letsencrypt/ "Ssl certificates for your azure website using Letsencrypt"
-[BlogReactNginxHTTP2]: /blog/react-router-and-nginx-over-http2/ "React Router and Nginx over HTTP/2"
+[blogletsencrypt]: /blog/ssl-certificate-for-your-azure-website-using-letsencrypt/ "Ssl certificates for your azure website using Letsencrypt"
+[blogreactnginxhttp2]: /blog/react-router-and-nginx-over-http2/ "React Router and Nginx over HTTP/2"

-In my [previous blog][BlogReactNginxHTTP2] I have already shown you how to use self-signed certificates in Nginx to use HTTP/2 features. I have also written a blog a long time ago on [how to get a Letsencrypt certificate for your Azure website][BlogLetsEncrypt].
+In my [previous blog][blogreactnginxhttp2] I have already shown you how to use self-signed certificates in Nginx to use HTTP/2 features. I have also written a blog a long time ago on [how to get a Letsencrypt certificate for your Azure website][blogletsencrypt].
```

**Comment by ChristianMurphy** [comment, marked as answer]
@marcofranssen this sounds like an [XY Problem](https://meta.stackexchange.com/questions/66377/what-is-the-xy-problem)?
reference links are case insensitive https://spec.commonmark.org/0.29/#example-174, uppercase vs lowercase doesn't matter for the link being resolved.
Furthermore Remark/MDX are [AST](https://en.wikipedia.org/wiki/Abstract_syntax_tree) based, they will preserve document structure but they don't preserve exact syntax as a [CST](https://en.wikipedia.org/wiki/Parse_tree) would.

**Reply by marcofranssen** [reply]
Right. But why are these links being lowercased by the `prettier.format` line? It would be great if this could be configured. It only happens when prettier kicks in, but skipping prettier results in other cleanup not happening. All the other parts are also using an AST in those use parts, they are not transforming those links. According to prettier they are using `remark-parse` https://prettier.io/docs/en/options.html#parser

**Reply by ChristianMurphy** [reply]
Again, remark is an AST, it doesn't preserve the casing, in the following example note that the reference identifiers are all lowercase https://astexplorer.net/#/gist/c268569a957949fa3f300914625219db/74e51ce678c64d467070ecc7d7c646d572673b7e

> It would be great if this could be configured

Keeping non-structural features is a non-goal for remark, AST's intentionally abstract away syntax details.
There has been discussion that https://github.com/micromark/micromark could produce a CST.
It hasn't been implemented yet.

> But why are these links being lowercased by the `prettier.format` line?

That would be a question for prettier.

**Reply by wooorm** [reply]
This is a decision made by Prettier: as the uppercase characters are unneeded, they decide to lowercase them.

Btw: mdast *does* include information on the original value in a label: at the `label` field: https://github.com/syntax-tree/mdast#association.

**Reply by ChristianMurphy** [reply]
looks like this has been cross-posted as https://github.com/prettier/prettier/issues/10007

**Reply by marcofranssen** [reply]
Correct cross posted here, when I figured prettier is using `remark-parse`. Will try to check with the prettier team if there is a way to keep existing labels as opposed to lowercasing all of them.
